From 8e8a824533643de4a4a221358a8c5ae7d4e9e474 Mon Sep 17 00:00:00 2001
From: Ameya Palande <ameya.palande@nokia.com>
Date: Wed, 28 Jul 2010 14:41:24 +0300
Subject: [PATCH] unknown hz value fix

Fixes linux version detection. This relied on a side-effect of the elf loader
and previously hoped that constructors would be called in a certain order.
That behavior was undefined and hence unreliable. For newer kernels, an elf
note is now checked to extract HZ tick count.

Original author: <david.sugar@canonical.com>
Backported to meego by: Ameya Palande <ameya.palande@nokia.com>
---
 proc/sysinfo.c |    1 +
 proc/version.c |    5 ++---
 proc/version.h |    1 +
 3 files changed, 4 insertions(+), 3 deletions(-)

diff --git a/proc/sysinfo.c b/proc/sysinfo.c
index 0ad7fba..793829f 100644
--- a/proc/sysinfo.c
+++ b/proc/sysinfo.c
@@ -212,6 +212,7 @@ static int check_for_privs(void){
 static void init_libproc(void) __attribute__((constructor));
 static void init_libproc(void){
   have_privs = check_for_privs();
+  init_Linux_version(); // make sure we have version before continuing...
   // ought to count CPUs in /proc/stat instead of relying
   // on glibc, which foolishly tries to parse /proc/cpuinfo
   //
diff --git a/proc/version.c b/proc/version.c
index 391b9bb..2049e15 100644
--- a/proc/version.c
+++ b/proc/version.c
@@ -35,11 +35,10 @@ void display_version(void) {
 
 int linux_version_code;
 
-static void init_Linux_version(void) __attribute__((constructor));
-static void init_Linux_version(void) {
+void init_Linux_version(void) {
     static struct utsname uts;
     int x = 0, y = 0, z = 0;	/* cleared in case sscanf() < 3 */
-    
+
     if (uname(&uts) == -1)	/* failure implies impending death */
 	exit(1);
     if (sscanf(uts.release, "%d.%d.%d", &x, &y, &z) < 3)
diff --git a/proc/version.h b/proc/version.h
index 492d251..28c9b49 100644
--- a/proc/version.h
+++ b/proc/version.h
@@ -14,6 +14,7 @@
 
 EXTERN_C_BEGIN
 
+extern void init_Linux_version(void);	/* initialize linux version */
 extern void display_version(void);	/* display suite version */
 extern const char procps_version[];		/* global buf for suite version */
 extern const char procps_number_version[];	/* global buf for suite number version */
-- 
1.7.0.4

